### PR TITLE
fix issue with the addAllPayloads

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 1]
```diff
@@ -404,6 +403,34 @@ struct PopAwareForkResolutionComparator {
...
 
+  void removeAllPayloads(protected_index_t& index,
+                         const std::vector<protected_payloads_t>& payloads) {
+    for (size_t i = 0, size = payloads.size(); i < size; i++) {
+      auto& p = payloads[i];
+      removePayloads(index, p);
+    }
+  }
+
+  bool proccedAllPayloads(protected_index_t& index,
```
Comment (blackyblack) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:414]: proceed

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 2]
```diff
@@ -404,6 +403,34 @@ struct PopAwareForkResolutionComparator {
     return true;
   }
 
+  void removeAllPayloads(protected_index_t& index,
+                         const std::vector<protected_payloads_t>& payloads) {
+    for (size_t i = 0, size = payloads.size(); i < size; i++) {
+      auto& p = payloads[i];
+      removePayloads(index, p);
+    }
```
Comment (blackyblack) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:411]: for (const auto &p : payloads) {
...
}

File: src/blockchain/pop/vbk_block_tree.cpp [hunk 3]
```diff
@@ -93,11 +93,8 @@ bool VbkBlockTree::acceptBlock(const VbkBlock& block,
     }
   }
 
-  // then, do stateful validation of payloads
-  if (!payloads.empty() && !cmp_.addAllPayloads(*index, payloads, state)) {
-    return state.Invalid("VbkTree::acceptBlock",
-                         "vbk-invalid-pop-" + state.GetRejectReason(),
-                         state.GetDebugMessage());
+  if (!cmp_.proccedAllPayloads(*index, payloads, state)) {
```
Comment (blackyblack) [inline src/blockchain/pop/vbk_block_tree.cpp:96]: proceed

File: include/veriblock/blockchain/block_index.hpp [hunk 4]
```diff
@@ -2,6 +2,7 @@
 #define ALT_INTEGRATION_INCLUDE_VERIBLOCK_BLOCKCHAIN_BLOCK_INDEX_HPP_
 
 #include <memory>
+#include <stack>
```
Comment (Warchant) [inline include/veriblock/blockchain/block_index.hpp:5]:
```suggestion
```

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 5]
```diff
@@ -386,8 +386,7 @@ struct PopAwareForkResolutionComparator {
       }
 
       // then, check if endorsement is valid
-      if (!sm.addPayloads(p, state)) {
-        removeContextFromBlockIndex(index, p);
+      if (!endorsementValidation(index, p, temp, protectedParams_, state)) {
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:389]:
```suggestion
      if (!checkEndorsement(index, p, temp, protectedParams_, state)) {
```

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 6]
```diff
@@ -404,6 +403,34 @@ struct PopAwareForkResolutionComparator {
...
+    for (size_t i = 0, size = payloads.size(); i < size; i++) {
+      auto& p = payloads[i];
+      removePayloads(index, p);
+    }
+  }
+
+  bool proccedAllPayloads(protected_index_t& index,
+                          const std::vector<protected_payloads_t>& payloads,
+                          ValidationState& state) {
+    if (!payloads.empty() && !tryValidateWithResources(
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:417]: move `tryValidateWithResources` inside of `addAllPayloads`

File: include/veriblock/blockchain/pop/pop_utils.hpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+#include "veriblock/blockchain/chain.hpp"
+#include "veriblock/validation_state.hpp"
+
+namespace altintegration {
+
+template <typename ProtectingBlockTree,
+          typename ProtectedIndex,
+          typename ProtectedChainParams>
+bool endorsementValidation(ProtectedIndex& index,
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/pop_utils.hpp:14]: function is named `endorsementValidation` but it also adds endorsement to BlockIndex... Maybe call it `checkAndAddEndorsement`?

File: include/veriblock/blockchain/pop/pop_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+      while (std::remove_if(endorsements.begin(),
+                            endorsements.end(),
+                            [&endorsement](endorsement_t* e) -> bool {
+                              // remove nullptrs and our given endorsement
+                              return !e || endorsement.get() == e;
+                            }) != endorsements.end())
+        ;
+
+      // remove from 'containing endorsements'
+      index.containingEndorsements.erase(endorsementit);
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/pop_utils.hpp:99]: this removal looks weird.
```
std::remove_if
```
does not remove elements... until you call `erase`
should be 
```
auto newEnd = std::remove_if(endorsements.begin(),
                            endorsements.end(),
                            [&endorsement](endorsement_t* e) -> bool {
                              // remove nullptrs and our given endorsement
                              return !e || endorsement.get() == e;
                            }), endorsements.end());
endorsements.erase(newEnd);
```

File: src/blockchain/pop/vbk_block_tree.cpp [hunk 9]
```diff
@@ -147,9 +143,9 @@ void addContextToBlockIndex(BlockIndex<VbkBlock>& index,
   auto add = [&](const BtcBlock& b) {
     auto hash = b.getHash();
     // filter context: add only blocks that are unknown and not in current 'ctx'
-    if (!tree.getBlockIndex(hash) && !set.count(hash)) {
+    if (!tree.getBlockIndex(hash) /*&& !set.count(hash)*/) {
       ctx.push_back(b);
-      set.insert(hash);
+      // set.insert(hash);
```
Comment (Warchant) [inline src/blockchain/pop/vbk_block_tree.cpp:148]: uncomment

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 10]
```diff
@@ -343,6 +343,8 @@ struct PopAwareForkResolutionComparator {
   const ProtectingBlockTree& getProtectingBlockTree() const { return tree_; }
   const protected_index_t* getIndex() const { return index_; }
 
+  // this function does not clear BlokcIndex from the added payloads (it is
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:346]:
```suggestion
  // this function does not clear BlockIndex from the added payloads (it is
```

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 11]
```diff
@@ -343,6 +343,8 @@ struct PopAwareForkResolutionComparator {
...
+  // doing by removePayloads). Better to use proccedAllPayloads() method
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:347]:
```suggestion
  // doing by removePayloads). Better to use processAllPayloads() method
```

**Pull request merged**